### ISSUE TITLE
File header comment sniff should allow plugin/theme header

### ISSUE THREAD

**Issue by felixarntz** [WordPress-Coding-Standards/WordPress-Coding-Standards#1070]
I see the following errors when sniffing against my plugin's main file:

* You must use `/**` style comments for a file comment
* Empty line required before block comment

This is shown since the first lines of that file mark the plugin header (for which I use `/*`). I even have the actual file comment, but it's right below the plugin header comment. Is there a recommended way I should fix that, or is that actually a problem in the sniff that needs to be adjusted for these cases?

**Comment by grappler** [comment]
@felixarntz Have you tried the second example of a plugin header?

https://codex.wordpress.org/File_Header#Plugin_File_Header_Example

Could you share your file header?

**Comment by JDGrimes** [comment]
The file header has to be the first thing in the file, so you should either switch or combine them. I've done both in different plugins.

**Comment by GaryJones** [comment]
You can combine the plugin headers into a DocBlock:

Here's my plugin boilerplate:

```php
/**
 * Plugin Name
 *
 * @package      Gamajo\PluginSlug
 * @author       Gary Jones
 * @copyright    2017 Gamajo
 * @license      GPL-2.0+
 *
 * @wordpress-plugin
 * Plugin Name:       Plugin Boilerplate
 * Plugin URI:        https://github.com/garyjones/...
 * Description:       ...
 * Version:           0.1.0
 * Author:            Gary Jones
 * Author URI:        https://gamajo.com
 * Text Domain:       plugin-slug
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * GitHub Plugin URI: https://github.com/garyjones/...
 */
```

The `@wordpress-plugin` annotation/tag is what I initiated several years ago, after talks with Mike van Riel, regarding custom tag names in DocBlocks. It helps delimit the rest of the DocBlock from the plugin headers.

In theory, phpDocumentor, the WP Parser, or others, could add in custom parsing for that block of plugin headers, to split it up into individual lines/values.

There are several WP plugin boilerplate projects which use it, so while it's not ubiquitous, it does meet the needs of WP, WPCS, PSR-5 and phpDocumentor .

**Comment by felixarntz** [comment]
Thanks all, I didn't know that was a valid way to handle the plugin header. Changed it, and works like a charm! 💯

**Comment by paulcanning** [comment]
Despite using the DocBlock method, I still get a warning for 'Missing file doc comment' at the top of the file?

**Comment by JDGrimes** [comment]
@paulcanning Can you paste/link to an example? The comment needs to be the very first thing in the file, no statements can come before it.

**Comment by paulcanning** [comment]
```
<?php
/**
 * My Plugin
 *
 * @package    MyPlugin
 * @author      Paul Canning
 * @copyright   2017
 * @license     GPL-2.0+
 *
 * @wordpress-plugin
 * Plugin Name: My Plugin
 * Description: My custom plugin.
 * Version:     1.0.0
 * Author:      Paul Canning
 * Text Domain: myplugin
 * License:     GPL-2.0+
 * License URI: http://www.gnu.org/licenses/gpl-2.0.txt
 */

require_once( ABSPATH . 'wp-admin/includes/plugin.php' );

...
```

**Comment by paulcanning** [comment]
Error is highlighting the very top line, `<?php`

**Comment by JDGrimes** [comment]
This happens whenever the first doc comment in a file is followed by a documentable element that doesn't have its own doc comment (for example, a class or function). When this happens, the doc comment is assumed to be associated with that element rather than it being the doc comment for the file. 

In this case, `include` and `require` statements are actually also counted as documentable elements, and therefore the sniff is assuming that the comment belongs to the `require_once` statement rather than being the file header. Adding a comment (even just `// for phpcs`) above the `require_once` statement will make the error go away.

**Comment by paulcanning** [comment]
Ah, excellent, seems to work, thanks!

**Comment by limitcracker** [comment]
> You can combine the plugin headers into a DocBlock:
> 
> Here's the file-level DocBlock from my plugin boilerplate:
> 
> ```
> /**
>  * Plugin Name
>  *
>  * @package      Gamajo\PluginSlug
>  * @author       Gary Jones
>  * @copyright    2017 Gamajo
>  * @license      GPL-2.0+
>  *
>  * @wordpress-plugin
>  * Plugin Name:       Plugin Boilerplate
>  * Plugin URI:        https://github.com/garyjones/...
>  * Description:       ...
>  * Version:           0.1.0
>  * Author:            Gary Jones
>  * Author URI:        https://gamajo.com
>  * Text Domain:       plugin-slug
>  * License:           GPL-2.0+
>  * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
>  * GitHub Plugin URI: https://github.com/garyjones/...
>  */
> ```
> 
> The `@wordpress-plugin` annotation/tag is what I initiated several _years_ ago, after talks with Mike van Riel, regarding custom tag names in DocBlocks. It helps delimit the rest of the DocBlock from the plugin headers.
> 
> Without it (and with everything else kept in the same order), phpDocumentor and others will complain that descriptive text can't come after tags.
> 
> In theory, phpDocumentor, the WP Parser, or others, could add in custom parsing for that block of plugin headers, to split it up into individual lines/values.
> 
> There are several WP plugin boilerplate projects which use it, so while it's not ubiquitous, it does meet the needs of WP, WPCS, PSR-5 and phpDocumentor.
> 
> > Have you tried the second example of a plugin header?
> 
> I [added](https://codex.wordpress.org/index.php?title=File_Header&action=historysubmit&diff=155493&oldid=151690) that to the wiki :-)

hi. thanks for the solution. will this formatting be accepted if submitted to WP Plugin Repos at some point?

**Comment by GaryJones** [comment]
> will this formatting be accepted if submitted to WP Plugin Repos at some point?

Yes - there's no issue about how you format your DocBlock / plugin headers.

Here's a [very old plugin of mine](https://plugins.trac.wordpress.org/browser/diagnosis/trunk/diagnosis.php), showing the combined usage.
